features/steps/thens/values.py

```diff
@@ -1,86 +1,102 @@
 import csv
-import errors as err
 import ifcopenshell
 import os
 
-from behave import *
+from behave import register_type
 from pathlib import Path
-from utils import ifc, misc
 
+from validation_handling import gherkin_ifc, StepResult, handle_errors
 
-@then("The value must {constraint}")
-@then("The values must {constraint}")
-@then('At least "{num:d}" value must {constraint}')
-@then('At least "{num:d}" values must {constraint}')
-@err.handle_errors
-def step_impl(context, constraint, num=None):
-    errors = []
+from parse_type import TypeBuilder
+register_type(unique_or_identical=TypeBuilder.make_enum(dict(map(lambda x: (x, x), ("be unique", "be identical"))))) # todo @gh remove 'be' from enum values
+register_type(value_or_type=TypeBuilder.make_enum(dict(map(lambda x: (x, x), ("value", "type"))))) # todo @gh remove 'be' from enum values
+
+
+@gherkin_ifc.step("The value must be in '{csv_file}.csv'")
+@gherkin_ifc.step("The values must be in '{csv_file}.csv'")
+def step_impl(context, inst, csv_file):
+    if not inst:
+        return []
+
+    dirname = os.path.dirname(__file__)
+    filename = Path(dirname).parent.parent / "resources" / f"{csv_file}.csv"
+    valid_values = [row[0] for row in csv.reader(open(filename))]
+    invalid_values = [value for value in inst if value not in valid_values]
+    for value in invalid_values:
+        yield StepResult(expected = f"Value in {csv_file}.csv", observed = value)
+
+    return []
+
+
+@gherkin_ifc.step('At least "{num:d}" value must {constraint}')
+@gherkin_ifc.step('At least "{num:d}" values must {constraint}')
+def step_impl(context, inst, constraint, num):
+    stack_tree = list(
+        filter(None, list(map(lambda layer: layer.get('instances'), context._stack))))
+
+    if num is not None:
+        values = list(map(lambda s: s.strip('"'), constraint.split(' or ')))
+
+        if stack_tree:
+            num_valid = 0
+            for i in range(len(stack_tree[0])):
+                path = [l[i] for l in stack_tree]
+                if path[0] in values:
+                    num_valid += 1
+            if num is not None and num_valid < num:
+                yield StepResult(expected = constraint, observed = f"Not {constraint}")
```

Comment:
This is probably nitpicking, but I have some objections to this:

- The expected value is not `constraint`, but `num` times `constraint`. It could be that `costraint` is actually encountered, but not enough times. One thing we can do is encode this a bit richer like `expected={"op": "count", "value": constraint, "num": num}`. Not saying it's the best idea, it's just a decision as a team how precise we want to be.

- For observed:

    - I don't really like flattening the output to strings, it would make it impossible for example to provide translations for the term Not. We could do `observed={"predicate": "not", "value": ...}`
    - But actually, this doesn't add information. Given the fact that an error is reported, one already knows that the value does not match the constraint. One solution I think could be to just leave it observed=None, if it's too much trouble.
    - Otherwise I would rather just pass the values from the stack. Maybe `observed = [l[0] for l in stack_tree]` I don't really remember the details of this. We should probably document this.